sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVExprUtils.scala

```diff
@@ -35,22 +35,11 @@ object CSVExprUtils {
     }
   }
 
-  def skipComments(iter: Iterator[String], options: CSVOptions): Iterator[String] = {
-    if (options.isCommentSet) {
-      val commentPrefix = options.comment.toString
-      iter.dropWhile { line =>
```

Comment:
Cannot replace this because this drops and skip filtering vs `filterCommentAndEmpty` entirely filter all iterator. The difference is necessary

Reply:
Could you help me understand why the behavior would change when switching the functions? `skipComments` is only used in `extractHeader(iter: Iterator[String], options: CSVOptions)` where it doesn't make a difference if `filter` or `dropWhile` is used. Is this an optimisation?

Reply:
Because with `DataFrameReader.csv(Dataset[String])` (filtering) you don't know where is the first block when you apply this function. But when you read it from file, you can know the start of the file - in this case we can drop (by partition number). The latter is more correct.